### PR TITLE
[dx12] aligned buffer-image copies

### LINKED ISSUE #1843: Fix subresource footprint calculation in D3D12
@zakorgy wrote:
> When creating the D3D12_SUBRESOURCE_FOOTPRINT the height variable should take the buffer offset in account. Because if the offset is not zero, the debug layer shows an error, that we want to write more bytes than the buffer size.  A possible solution which worked for us, is to replace this [line](https://github.com/gfx-rs/gfx/blob/8d0ef5ba96bb4a2ecc5671975677323f11773dc5/src/backend/dx12/src/command.rs#L1296) with
```rust
let height = height as u32 - (region.buffer_offset as u32 / row_pitch);
```

### PR DESCRIPTION
**Pull request by kvark**
Replaces  #1877
Fixes #1843

PR checklist:
- [ ] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [x] tested examples with the following backends:

### CODE REVIEW

File: src/backend/dx12/src/command.rs [hunk 1]
```diff
@@ -602,6 +616,148 @@ impl CommandBuffer {
...
+                            .. buf_offset
+                        },
+                        copy_extent: image::Extent {
+                            width: r.image_extent.width - half,
+                            .. r.image_extent
+                        },
+                    });
+                }
+            } else {
+                // worst case: row by row copy
```
Comment (msiglreith) [inline src/backend/dx12/src/command.rs]: Could you explain the idea here?
Imo the worst case scenario here is a image laid out in memory like this, considering row pitch alignment and data alignment:
```
-----------
----------x
xxxxxxxxxoo
ooooooooyyy
yyyyyyyzzzz
zzzz ...
```
The resulting image should be
```
xxxxxxxxxx
oooooooooo
yyyyyyyyyy
zzzzzzzzzz
...
```

File: src/backend/dx12/src/command.rs [hunk 2]
```diff
@@ -602,6 +616,148 @@ impl CommandBuffer {
...
+                for z in 0 .. r.image_extent.depth {
+                    for y in 0 .. r.image_extent.height {
+                        let row_offset = layer_offset + z as u64 * slice_pitch as u64 + y as u64 * row_pitch as u64;
+                        let footprint_offset = row_offset & !(d3d12::D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT as u64 - 1);
+                        let gap = (row_offset - footprint_offset) as u32;
+                        copies.push(Copy {
+                            footprint_offset,
+                            footprint: image::Extent {
+                                width: r.image_extent.width + gap,
+                                height: 1,
```
Comment (msiglreith) [inline src/backend/dx12/src/command.rs]: Should check block dimension (1, 1) here

**Comment by kvark** [comment]
@msiglreith thanks for spotting the problems! should be addressed now, PTAL

File: src/backend/dx12/src/command.rs [hunk 3]
```diff
@@ -602,6 +616,182 @@ impl CommandBuffer {
...
+            r.buffer_height
+        };
+        let row_pitch = div(buffer_width, image.block_dim.0 as _) * image.bytes_per_block as u32;
+        let slice_pitch = div(buffer_height, image.block_dim.1 as _) * row_pitch;
+        let is_pitch_aligned = row_pitch % d3d12::D3D12_TEXTURE_DATA_PITCH_ALIGNMENT == 0;
+
+        for layer in r.image_layers.layers.clone() {
+            let img_subresource = image
+                .calc_subresource(r.image_layers.level as _, layer as _, 0);
+            let layer_offset = r.buffer_offset as u64 + (layer as u32 * slice_pitch * r.image_extent.depth) as u64;
```
Comment (msiglreith) [inline src/backend/dx12/src/command.rs]: `* r.image_extent.depth` probably not needed

File: src/backend/dx12/src/command.rs [hunk 4]
```diff
@@ -602,6 +616,182 @@ impl CommandBuffer {
...
+                        });
+
+                        // and if it crosses a pitch alignment - we copy the rest separately
+                        if cut_width == r.image_extent.width {
+                            continue;
+                        }
+                        let leftover = r.image_extent.width - cut_width;
+
+                        copies.push(Copy {
+                            footprint_offset: next_aligned_offset,
```
Comment (msiglreith) [inline src/backend/dx12/src/command.rs]: This looks somewhat suspicious to me as `D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT` is twice as large as the required row pitch.

**Comment by kvark** [comment]
Noted. Thanks for the review!
Bors r=msiglreith

**Pull request merged**